foresight/utils/test_wrapper_utils.py

```diff
@@ -0,0 +1,266 @@
+            try:
+                root_span.finish(f_time=execution_context.finish_timestamp)
+            except Exception:
+                # TODO: handle root span finish errors
+                pass
+            finally:
+                scope.close()
+        except Exception as err:
+            logger.error("test wrapper finish trace error: {}".format(err))
+            pass
+
+
+    def start_invocation(self, execution_context):
+        """Calling after thundra invocation pluging has been called by execute_hook. Create trace span.
+
+        Args:
+            execution_context (TestSuiteExecutionContext | TestSuiteExecutionContext): Execution context
+        """
+        try:
+            execution_context.invocation_data = wrapper_utils.create_invocation_data(self.plugin_context, execution_context)
+        except Exception as err:
+            logger.error("test wrapper start invocation error: {}".format(err))
+            pass
+
+
+    def finish_invocation(self, execution_context):
+        """Finish invocation.
+
+        Args:
+            execution_context (TestSuiteExecutionContext | TestSuiteExecutionContext): Execution context
+        """
+        try:
+            wrapper_utils.finish_invocation(execution_context)
+        except Exception as err:
+            logger.error("finish invocation error: {}".format(err))
+            pass
+
+
+    def before_test_process(self, execution_context):
+        """Invoke trace and invocation plugings before each test suite and test case.
+
+        Args:
+            execution_context (TestSuiteExecutionContext | TestCaseExecutionContext): Thundra test execution context
+        """
+        try:
+            self.execute_hook("before:invocation", execution_context)
+        except Exception as err:
+            logger.error("test wrapper before test process: {}".format(err))
+            pass
+
+
+    def after_test_process(self, execution_context):
+        """Send collected data about test suit or test case after each finish.
+
+        Args:
+            execution_context (TestSuiteExecutionContext | TestCaseExecutionContext): Thundra test execution context
+        """
+        try:
+            self.prepare_and_send_reports_async(execution_context)
+        except Exception as err:
+            logger.error("test wrapper after test process error: {}".format(err))
+            pass
+
+
+    def send_test_run_data(self, test_run_event):
+        """Send test run data. test_run_monitoring data has been updated by application info because
+         reporter has been used tih info whilst preparing composite data. 
+
+        Args:
+            test_run_event (TestRunStart, TestRunStatus, TestRunFinish): Tests event data
+        """
+        try:
+            test_run_monitoring_data = test_run_event.get_monitoring_data()
+            app_info = self.application_info_provider.get_application_info()
+            test_run_monitoring_data["data"].update(app_info)
```

Comment:
why we need to inject application info here?

Reply:
It is because of reporter flow. Reporter check application info by using _common_fields_list. Otherwise, http error occurs.

Reply:
I have proposed another solution to get rid of this logic: https://github.com/thundra-io/thundra-agent-python/pull/171/files#r740743321

Reply:
done.